Migration to GDAL 3

dcdenu4 opened the pull request:
This PR migrates Pygeoprocessing to GDAL 3 and drops GDAL 2. 

GDAL 3 comes with some major changes which can be viewed here: 
- https://trac.osgeo.org/gdal/wiki/Release/3.0.0-News
- https://gdalbarn.com/#why
- https://trac.osgeo.org/gdal/wiki/rfc73_proj6_wkt2_srsbarn

The main points from the barn raising:

> “Coordinate systems in GDAL, PROJ, and libgeotiff are missing modern capabilities and need a thorough refactoring:
The dreaded ad hoc CSV databases in PROJ_LIB and GDAL_DATA are frustrating for users, pose challenges for developers, and impede interoperability of definitions.
GDAL and PROJ are missing OGC WKT2 support.
PROJ 5.0+ no longer requires datum transformation pivots through WGS84, which can introduce errors of up to 2m, but the rest of the tools do not take advantage of it.”

The major changes that actively affect Pygeoprocessing is the datum transformation through WGS84. Coordinate transformations now expect lat,lon instead of lon,lat. The recommendation is to use `osr.SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER)` which will allows for GIS friendly usage.

The proposed solution is to check Spatial References when creating coordinate transformations and if the spatial reference is unprojected, apply the migration Axis recommendation.

One hard coded test was updated to reflect slight variance in extents.

dcdenu4 committed:
Update issue #20 adding Python 3.8 to test matrix

dcdenu4 committed:
Update issue #16 Support for GDAL 3

Adding support to handle GDAL 3. This effectively ends support for
GDAL 2. When using coordinate transformations now setting Axis
properties to be 'friendly' which means to expect lon,lat as opposed to
the now standard of lat,lon. This was recommonended in the migration
guide from GDAL.

Changing the values of a hardcoded test slightly as spatial references
are slightly different in GDAL3, so minor extent differences are
expected in some instances.

dcdenu4 committed:
Merge branch 'master' into bugfix/python-38-support

dcdenu4 committed:
Issue #16 adding gdal 3 remark in history.

Adding python 3.8 to setup.py for support as all GDAL 3 testing has been
done using python 3.8

dcdenu4 committed:
Issue #16 Forcing gdal 3 or higher

dcdenu4 committed:
Issue #16 Removing Python 3.8 support

richpsharp commented:
Hey Doug, just FYI can you include me on pull requests on PyGeoprocessing in the future?

richpsharp commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1630,8 +1630,18 @@ def reproject_vector(
+    if not base_sr.IsProjected():
+        base_sr.SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER)
+    if not target_sr.IsProjected():
+        target_sr.SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER)
```
~~If I understand the issue correctly, in this function everything is contained inside of `osr.Geometry` objects which should be internally consistent with whatever the axis order is for whatever the projection is. We shouldn't encounter an x vs. y vs lat vs long because it's not exposed at this level. Is that possible? Or did you make this change because of an issue in InVEST or something?~~

richpsharp replied:
~~p.s. I'm not totally sure about this, but I'd want to make sure we know why we're setting the axis mapping strategy. If there's a failed test or something that's obvious, otherwise I keep thinking "how do we know if it's lat/lng or x/y". In the bounding box routine below we really don't know. But here all the geometry should know and `ogr` should be internally consistent?~~

richpsharp replied:
RE: a thread we had on slack, yes this is necessary, but not necessary to have the `if` statements in there. And I have one other overall PR comment related to this I'll put up top.

dcdenu4 replied:
Yep, totally agree.

richpsharp commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -2722,10 +2732,23 @@ def transform_bounding_box(
+    if not base_ref.IsProjected():
+        base_ref.SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER)
+    if not target_ref.IsProjected():
+        target_ref.SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER)
```
Reading through https://trac.osgeo.org/gdal/wiki/rfc73_proj6_wkt2_srsbarn#Axisorderissues, I think the test for not [...].IsProjected(): might not be necessary, and in some cases wrong. Like some kind of exotic projected transformation that swaps its coordinates. From what I understand in the case here where we're always assuming x/y order, we should assume the osr.OAMS_TRADITIONAL_GIS_ORDER for both transforms. Is that right?

dcdenu4 replied:
Hey Rich, I just spent a significant amount of time in the WKT2 docs, PROJ 6 docs, and GDAL 3 Docs. I _think_ I'm starting to develop a deeper understanding. I think the basics are as follows:

- GDAL 3 conforms to the authority defining coordinate reference systems (CRS). 
- For Geographic CRS (WGS84), this means no longer ignoring Axis Order but abiding by LAT,LON order
- For Projected CRS (UTM), most EPSG in the registry use Easting, Northing order
- We think of things in general as **x**,**y** because it is convenient, but now a `Transform` expects Axis Order, where  if `long=x` and `lat=y`, GDAL 3 would take `Transform(y, x)`
- Thus when transforming from Projected CRS to Geographic CRS, `Transform(x,y)` will return `y,x` (Lat,Lon)
- Thus when transforming from Geographic CRS to Projected CRS, `Transform(y,x)` will return `x,y` (Easting, Northing)
- Thus when transforming from Geographic CRS to Geographic CRS, `Transform(y,x)` will return `y,x` (Lat,Lon)
- Thus when transforming from Projected CRS to Projected CRS, `Transform(x,y)` will return `x,y` (Easting, Northing)

**note:** Most projected CRS use Easting first, Northing second convention, however, some defined in the EPSG registry use the reverse convention.

So, I think you're correct in that `not.____.IsProjected()` is probably wrong, but `_____.IsGeographic()` _could_ be OK. Basically that Axis Order setting `SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER)` is saying, if we've got a Geographic CRS, use Lon,Lat like the good old days (since we think of `lon,lat` as `x,y`. We certainly can, but maybe should, **NOT** do that. 

Perhaps a better solution is to create a mapping of Axis Order to Geographic (Lat,Lon) and Projected (Easting, Northing), such that when handling the extents and bounding box points we can correctly return _x_, _y_.

richpsharp replied:
I think it's always okay to use `srs.SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER)` in this case. It'll only "swap" x/y if it's in a coordinate system that needs it. UTM projections are not affected. At least that's how I read the docs, and also I tried it manually and it behaved as expected. Is that okay?

dcdenu4 replied:
Yea, I think that's right!

richpsharp commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1630,8 +1630,18 @@ def reproject_vector(
+    if not target_sr.IsProjected():
+        target_sr.SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER)
+        
+    coord_trans = osr.CreateCoordinateTransformation(base_sr, target_sr)
```
But do should keep `CreateCoordinateTransformation` because that's a GDAL 3.0 thing!

richpsharp commented on HISTORY.rst:
```diff
@@ -8,6 +8,7 @@ Unreleased Changes
   pygeoprocessing source distribution.  This fixes an issue where files
   matching a variety of extensions anywhere in the pygeoprocessing directory
   might be included with the source distribution.
+* Adding GDAL 3 support and dropping GDAL 2 support.
```
Can you indicate that the only non-backwards compatible issues in GDAL 2 to GDAL 3 were the introduction of `osr.CreateCoordinateTransformation` (vs `CoordinateTransformation`) and the need to set `[srs_ref].osr.SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER)` because of https://trac.osgeo.org/gdal/wiki/rfc73_proj6_wkt2_srsbarn#Axisorderissues? I think this will only be an issue for people who manually construct

dcdenu4 replied:
I definitely will, but wanted to note that `osr.CoordinateTransformation` still works in GDAL 3, but since we're not supporting GDAL 2, can move on to the newer recommended function.

richpsharp reviewed (changes requested):
One last thing Doug, since most of these issues are all about lat vs. long or not, can you add test cases for transform_bounding_box and reproject_vector for testing transforms from utm->utm, wgs84->wgs84, wgs84->utm, and I think there's already one for utm->wgs84, but if not add that too?

dcdenu4 commented:
Hey @richpsharp, I just spent a significant amount of time in the WKT2 docs, PROJ 6 docs, and GDAL 3 Docs. I _think_ I'm starting to develop a deeper understanding. I think the basics are as follows:

- GDAL 3 conforms to the authority defining coordinate reference systems (CRS). 
- For Geographic CRS (WGS84), this means no longer ignoring Axis Order but abiding by LAT,LON order
- For Projected CRS (UTM), most EPSG in the registry use Easting, Northing order
- We think of things in general as **x**,**y** because it is convenient, but now a `Transform` expects Axis Order, where  if `long=x` and `lat=y`, GDAL 3 would take `Transform(y, x)`
- Thus when transforming from Projected CRS to Geographic CRS, `Transform(x,y)` will return `y,x` (Lat,Lon)
- Thus when transforming from Geographic CRS to Projected CRS, `Transform(y,x)` will return `x,y` (Easting, Northing)
- Thus when transforming from Geographic CRS to Geographic CRS, `Transform(y,x)` will return `y,x` (Lat,Lon)
- Thus when transforming from Projected CRS to Projected CRS, `Transform(x,y)` will return `x,y` (Easting, Northing)

**note:** Most projected CRS use Easting first, Northing second convention, however, some defined in the EPSG registry use the reverse convention.

So, I think you're correct in that `not.____.IsProjected()` is probably wrong, but `_____.IsGeographic()` _could_ be OK. Basically that Axis Order setting `SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER)` is saying, if we've got a Geographic CRS, use Lon,Lat like the good old days (since we think of `lon,lat` as `x,y`. We certainly can, but maybe should, **NOT** do that. 

Perhaps a better solution is to create a mapping of Axis Order to Geographic (Lat,Lon) and Projected (Easting, Northing), such that when handling the extents and bounding box points we can correctly return _x_, _y_.

phargogh committed:
Adding a workflow to build python wheels.

phargogh committed:
Correcting indentation level of steps.

phargogh committed:
Forgot a uses.

phargogh committed:
Trying a different build matrix format.

phargogh committed:
Trying another build matrix exclusion.

phargogh committed:
Correcting order.

phargogh committed:
Minor tweaks to config.

phargogh committed:
Disabling cibuildwheel for the time being.

phargogh committed:
Removing cibuildwheel entirely.

phargogh committed:
Renaming job.

phargogh committed:
Removing linux, manylinux will be better later.

phargogh committed:
Adding a source distribution. RE:#32

phargogh committed:
Installing deps for sdist, noting change in HISTORY. Fixes #32

phargogh committed:
Updating labels. RE:#32

phargogh committed:
Renaming workflow. RE:#32

phargogh committed:
Updating project link to github. RE:#33

phargogh committed:
Noting change in HISTORY. Fixes #33.

phargogh commented:
Hey @dcdenu4 I'm sorry I haven't had a chance to take a look at this closely (hoping to do so this afternoon!) but since this is kind of a forward-facing change, could you change the PR merge target into the `release/2.0` branch instead of `master`?  That will help us to continue to get bugfixes out on the 1.9.x release cycle and allow us to collect a few features for the 2.0 release in the near future.

Thanks!

dcdenu4 committed:
Issue #16 Adding more detailed info to history

dcdenu4 committed:
Issue #16 Fixing coodinatetransformation axis set

There is no need to check if the SRS is geographic or projected,
GDAL 3 appears to only swap when expected, like for lat/lon
to lon/lat for GIS friendly axis order.

Also removing the axis swapping in reproject_vector where the
geometries actually handle this internally.
Adding better comments.

richpsharp commented:
Hi @dcdenu4, another thing I'm hoping you can add: `pygeoprocessing` makes a lot of decisions about global state constants but exposes all of them to the user in one form or another. A big example of this is PyGeoprocessing's default to create GeoTIFFs and with LZW compressed bigtiff creation. That's almost always fine, but when it's not we don't want to build a separate version of `pygeoprocessing` (Peter did this once and me too). That's why we exposed the default as `raster_creation_options` in every `pygeoprocessing` function that creates a raster. 

Keeping with that convention, as part of this PR could you make a global default for the axis mapping strategy to `osr.OAMS_TRADITIONAL_GIS_ORDER` but expose that parameter in every function that touches any `pygeoprocessing` function that needs to do a transform? That'd be `reproject_vector` and `transform_bounding_box` because they do it directly, but also `align_and_resize_raster_stack` and `warp_raster` because they make a call to `transform_bounding_box`? And anywhere else you see that might be missing?

dcdenu4 committed:
Issue #16 Adding Axis Order to reproject_vector

This is needed because when the geometries are transformed to
the target SRS, by default GDAL 3 will now return those as
Lat,Lon, however they're handled as Lon,Lat. Setting this
has the transformed geometries returned as Lon,Lat if using a
geographic SRS

dcdenu4 committed:
Issue #16 Adding tests for Axis Ordering.

Adding tests to transform_bounding_box and reproject_vector functions
where osr.CreateCoordinateTransformation are handled. The tests for
transform bounding box use coordinates from a polygon made in QGIS
in the New England Coastal area for WGS84 and UTM 19N. The expected
results were varified in QGIS.

The test for reproject vector handle all possible transformation cases
in and out of WGS84 and Projected SRS. Specifically the latlon to latlon
test checks that the vectors are identical as well as the SRS is as
expected. To handle WGS84 I added a WGS84 reference to sampledata,
following the conventions for the other used SRS.

dcdenu4 committed:
Issue #16 removing testing.sampledata usage

Refactoring reproject_vector test functions to create their own vectors
for transforming.

dcdenu4 committed:
Issue #16 expose osr axis strategy

Like raster creation options, exposing OSR Axis Order Mapping Strategy
for creating coordinate transformations in def transform_bounding_box
and def reproject_vector.

dcdenu4 committed:
Merge pull request #35 from phargogh/bugfix/update-links-from-bitbucket-to-github

Bugfix/update links from bitbucket to GitHub

phargogh committed:
Merge branch 'master' into bugfix/build-wheels-on-github-actions

phargogh committed:
Updating checkout to v2. Fixes #36

phargogh committed:
Updating checkout to v2 and clarifying fetch depth. RE#32

phargogh committed:
Merge branch 'bugfix/build-wheels-on-github-actions' of github.com:phargogh/pygeoprocessing into bugfix/build-wheels-on-github-actions

dcdenu4 committed:
Merge pull request #38 from phargogh/bugfix/update-actions-checkout-to-v2

Updating checkout to v2. Fixes #36

dcdenu4 committed:
Merge branch 'master' into bugfix/python-38-support

dcdenu4 committed:
Issue #16 bumping ubuntu version to 18.04

Also adding conda-forge and anaconda to conda channels

dcdenu4 committed:
Issue #16 Fixing mistake of deleting lines

richpsharp commented:
Hi @dcdenu4 and @phargogh, can I ping this PR to keep it moving? I'm assuming it's held up on those failing tests right? But those tests are failing on Ubuntu specifically where there's an EPSG code that doesn't import from `pygeoprocessing.testing`. This is not a failure of PyGeoprocessing at all, but of an arbitrary import of an EPSG code that would have the same issue whether someone was using PyGeoprocessing or not. Since we're proposing removing `pygeoprocessing.testing` entirely, could we ignore those ubuntu test failures and proceed with this PR?

richpsharp committed:
Merge pull request #34 from phargogh/bugfix/build-wheels-on-github-actions

Bugfix/build wheels on GitHub actions

richpsharp commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -574,6 +578,9 @@ def align_and_resize_raster_stack(
             tuple/list as the second. Defaults to a GTiff driver tuple
             defined at geoprocessing.DEFAULT_GTIFF_CREATION_TUPLE_OPTIONS.
+        osr_axis_mapping_strategy (OSR object): The OSR Axis Mapping Strategy 
+            as the first element. Defaults to a GIS friendly Lon,Lat order 
```
It's not an OSR object, rather just an integer, can you change the type to (int)? Also the mention of "first element" sounds like it's a tuple. Could this read simply as:

```
osr_axis_mapping_strategy (int): OSR axis mapping strategy for `SpatialReference` objects. Defaults to 
    `geoprocessing.DEFAULT_OSR_AXIS_MAPPING_STRATEGY`. This parameter should not be changed unless you 
    know what you are doing.
```

phargogh replied:
I'd second this, with one *trivial* change, which is that our docstrings are RST, so inline literal code should be surrounded by double-backticks.

dcdenu4 replied:
Done in [2eba123](https://github.com/natcap/pygeoprocessing/pull/31/commits/2eba123c585913539377eeaa05ebccdf13990a80)

richpsharp commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1581,6 +1589,9 @@ def reproject_vector(
             ignored. If ``False`` no fields are copied into the new vector.
+        osr_axis_mapping_strategy (OSR object): The OSR Axis Mapping Strategy 
+            as the first element. Defaults to a GIS friendly Lon,Lat order 
+            defined at geoprocessing.DEFAULT_OSR_AXIS_MAPPING_STRATEGY.
```
Could you change the docstring to the one I proposed above?

dcdenu4 replied:
Done in [2eba123](https://github.com/natcap/pygeoprocessing/pull/31/commits/2eba123c585913539377eeaa05ebccdf13990a80)

richpsharp commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -503,7 +506,8 @@ def align_and_resize_raster_stack(
         vector_mask_options=None, gdal_warp_options=None,
-        raster_driver_creation_tuple=DEFAULT_GTIFF_CREATION_TUPLE_OPTIONS):
+        raster_driver_creation_tuple=DEFAULT_GTIFF_CREATION_TUPLE_OPTIONS,
+        osr_axis_mapping_strategy=DEFAULT_OSR_AXIS_MAPPING_STRATEGY):
```
👍

richpsharp commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -84,6 +84,9 @@
 }
 
+# GDAL 3 OSR Coordinate Transformation Axis Order strategy Lon,Lat
+DEFAULT_OSR_AXIS_MAPPING_STRATEGY = osr.OAMS_TRADITIONAL_GIS_ORDER
```
👍

phargogh replied:
Very nice ... consistent with the existing pygeoprocessing defaults!

richpsharp commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1629,9 +1640,18 @@ def reproject_vector(
+    # See https://trac.osgeo.org/gdal/wiki/rfc73_proj6_wkt2_srsbarn Axis order
+    # issues. SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER) swaps 
+    # the axis order, which will use Lon,Lat order for Geographic CRS, 
+    # but otherwise leaves Projected CRS alone. 
```
I think this comment isn't relevant here anymore since we pass in the axis mapping strategy as a parameter, but I think it may be helpful at the top of the module where you define the default. Would you mind moving it to the single default instance at the top?

dcdenu4 replied:
Done in [c9f0d6](https://github.com/natcap/pygeoprocessing/pull/31/commits/c9f0d66d3513cb8559367280e605b008d244f30c)

richpsharp commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1815,6 +1836,9 @@ def warp_raster(
             defined at geoprocessing.DEFAULT_GTIFF_CREATION_TUPLE_OPTIONS.
+        osr_axis_mapping_strategy (OSR object): The OSR Axis Mapping Strategy 
+            as the first element. Defaults to a GIS friendly Lon,Lat order 
+            defined at geoprocessing.DEFAULT_OSR_AXIS_MAPPING_STRATEGY.
```
Could you change the docstring to the one I proposed above?

dcdenu4 replied:
Done in [2eba123](https://github.com/natcap/pygeoprocessing/pull/31/commits/2eba123c585913539377eeaa05ebccdf13990a80)

richpsharp commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -2709,6 +2734,10 @@ def transform_bounding_box(
+            as the first element. Defaults to a GIS friendly Lon,Lat order 
+            defined at geoprocessing.DEFAULT_OSR_AXIS_MAPPING_STRATEGY.
+
 
```
Are there too many lines here w/r/t PEP 257?

richpsharp commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -2722,10 +2751,23 @@ def transform_bounding_box(
+    # See https://trac.osgeo.org/gdal/wiki/rfc73_proj6_wkt2_srsbarn Axis order
+    # issues. SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER) swaps 
+    # the axis order, which will use Lon,Lat order for Geographic CRS, 
+    # but otherwise leaves Projected CRS alone
```
Same comment as above

dcdenu4 replied:
Done in [c9f0d6](https://github.com/natcap/pygeoprocessing/pull/31/commits/c9f0d66d3513cb8559367280e605b008d244f30c)

richpsharp commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -2722,10 +2751,23 @@ def transform_bounding_box(
     def _transform_point(point):
         """Transform an (x,y) point tuple from base_ref to target_ref."""
+        # NOTE: In GDAL 3, if osr.OAMS_TRADITIONAL_GIS_ORDER not set then
+        # TransformPoint expects Lat,Lon and NOT Lon,Lat for Geographic CRS. 
```
This is only true if the projection system has the {2, 1} axis order like WGS84 has. Could you either remove this comment or point out something more general like "TransformPoint's axis order is dependent on the `SetAxisMappingStrategy` set above".

dcdenu4 replied:
removed in [779ec92](https://github.com/natcap/pygeoprocessing/pull/31/commits/779ec92c5b38e30880d6d986ccdb88ddf644cbf7)

richpsharp commented on src/pygeoprocessing/testing/sampledata.py:
```diff
@@ -113,6 +113,11 @@ def projection_wkt(epsg_id):
+    projection=projection_wkt(4326),
+    origin=(-123.71107369348861, 44.76009901797307),
+    pixel_size=lambda x: (x, -1. * x)
+)
```
Is this used anywhere? I tried to "ctrl+f" it and nothing turned up. If not, can you remove it?

dcdenu4 replied:
Good point! I think I added it when I was creating earlier tests before we talked about having the `testing` module go away. And then I refactored those tests!

dcdenu4 replied:
Removed in [03b39f](https://github.com/natcap/pygeoprocessing/pull/31/commits/03b39fec8d4b72c9b1315082b19c4c171f14c79c)

richpsharp commented on tests/test_geoprocessing.py:
```diff
@@ -285,8 +294,314 @@ def test_reproject_vector(self):
+        import pygeoprocessing
+        # Create polygon shapefile to reproject
+        base_srs = osr.SpatialReference()
+        base_srs.ImportFromEPSG(4326) # WGS84
```
Can you add a space for the comment re: "PEP257, two spaces before inline # comment"?

dcdenu4 replied:
I think I got pep fixes in [f2c337b](https://github.com/natcap/pygeoprocessing/pull/31/commits/f2c337bce218e8c7f7d922edf0eb32f4d962f6df)

richpsharp commented on tests/test_geoprocessing.py:
```diff
@@ -285,8 +294,314 @@ def test_reproject_vector(self):
+        import pygeoprocessing
+        # Create polygon shapefile to reproject
+        base_srs = osr.SpatialReference()
+        base_srs.ImportFromEPSG(4326) # WGS84
```
two space before #

richpsharp commented on tests/test_geoprocessing.py:
```diff
@@ -2466,6 +2712,101 @@ def test_transform_box(self):
+
+        # lat/lon bounds for Coast of New England from shapefile created in
+        # QGIS 3.10.2 running GDAL 3.0.3
+        bounding_box = [-72.3638439, 40.447948, -68.0041948, 43.1441579]
```
Cool

richpsharp commented on tests/test_geoprocessing.py:
```diff
@@ -2466,6 +2712,101 @@ def test_transform_box(self):
+        expected_result = [
+            214722.122449, 4477484.382162, 584444.275934, 4782318.029707]
+
+        self.assertIs(
+            numpy.testing.assert_allclose(
+                result, expected_result), None)
```
Ha!

richpsharp reviewed (changes requested):
Great @dcdenu4, I only have some comments on the docstrings and comments. I hope you don't mind moving those around a little bit! Let me know if anything seems unreasonable okay?

richpsharp commented:
> Hi @dcdenu4 and @phargogh, can I ping this PR to keep it moving? I'm assuming it's held up on those failing tests right? But those tests are failing on Ubuntu specifically where there's an EPSG code that doesn't import from `pygeoprocessing.testing`. This is not a failure of PyGeoprocessing at all, but of an arbitrary import of an EPSG code that would have the same issue whether someone was using PyGeoprocessing or not. Since we're proposing removing `pygeoprocessing.testing` entirely, could we ignore those ubuntu test failures and proceed with this PR?

Oops I guess this was actually waiting on me. I don't know how I'm supposed to have notice there were changes pushed but I just happened to look at another tab. Sorry it was waiting on me, but also what do you think about the failing ubuntu test case points I made above?

dcdenu4 commented:
@richpsharp Thanks for taking another look! Last Friday I was able to get the GDAL_DATA and PROJ_LIB environment variables properly setting with using the [**setup-miniconda**](https://github.com/goanpeca/setup-miniconda) Github Actions service as opposed to the **setup-conda** Github Actions service. The **setup-miniconda** version has more stars and is maybe better developed? I'm thinking it might be worth switching to use that resource if subsequent Windows and Mac tests don't mind. 

But, I'm going to focus on this PR today and I will ping you when more updates get pushed.

phargogh commented:
> Last Friday I was able to get the GDAL_DATA and PROJ_LIB environment variables properly setting with using the setup-miniconda Github Actions service as opposed to the setup-conda Github Actions service. The setup-miniconda version has more stars and is maybe better developed? I'm thinking it might be worth switching to use that resource if subsequent Windows and Mac tests don't mind.

Personally, I can't believe I didn't find the `setup-miniconda` action.  It looks better-developed, better-documented and better-supported than the `setup-conda` action, so I'd be very happy to use it if it works well!

phargogh commented on .github/workflows/pythonpackage.yml:
```diff
@@ -12,16 +12,20 @@ jobs:
+          fetch-depth: 0
+
     - name: setup-conda
       uses: s-weigand/setup-conda@v1.0.0
```
If `setup-miniconda` is preferable, should we be using it here as well?

dcdenu4 replied:
This is a great question and I have some updates about this. But I'm leaning towards Rich's comments that maybe that should be a separate issue and PR, because it's currently complicated. Complicated in that `setup-miniconda` works for `mac` and `linux` but now fails on `windows` because of how Miniconda is added to Windows Path.

phargogh replied:
Uffda.  Sounds great to have this be in a separate PR!

phargogh commented on tests/test_geoprocessing.py:
```diff
@@ -285,8 +294,314 @@ def test_reproject_vector(self):
+        polygon_a = shapely.geometry.Polygon([
+            (extents[0], extents[3]), (extents[0], extents[1]),
+            (extents[2], extents[1]), (extents[2], extents[3]),
+            (extents[0], extents[3])])
```
It looks like this is constructing a box using the `extents` noted above.  What you have here totally works, but I just wanted to  mention that `shapely` has a convenience function [`shapely.geometry.box`](https://shapely.readthedocs.io/en/latest/manual.html#shapely.geometry.box) that will do the same thing in a slightly more compact notation.

dcdenu4 replied:
Great suggestion done in commit: [f2c337b](https://github.com/natcap/pygeoprocessing/pull/31/commits/f2c337bce218e8c7f7d922edf0eb32f4d962f6df)

phargogh commented on tests/test_geoprocessing.py:
```diff
@@ -285,8 +294,314 @@ def test_reproject_vector(self):
+        test_driver = None
+
+        target_reference = osr.SpatialReference()
+        target_reference.ImportFromEPSG(3157) # UTM zone 10N
```
There's another `PEP8` instance here for a comment.  There might be others here in this file as well ... could you try running `pycodestyle` (`pip install pycodestyle`) to check?

dcdenu4 replied:
Done here: [f2c337b](https://github.com/natcap/pygeoprocessing/pull/31/commits/f2c337bce218e8c7f7d922edf0eb32f4d962f6df)
Thanks for the `pycodestyle` tool!

phargogh reviewed (commented):
Hey @dcdenu4 I just went through this and added a few small comments.  Since @richpsharp has been much more involved in this in this package, I'll defer to him on the final PR approval.

dcdenu4 committed:
adding clearer blurb for osr_axis_mapping_strategy

dcdenu4 committed:
Moving gdal 3 spatial ref comment to top

dcdenu4 committed:
removing unneeded comment

dcdenu4 committed:
removing srs defn from sampledata, no longer used

dcdenu4 committed:
revert ubuntu os

dcdenu4 committed:
Merge branch 'master' into bugfix/python-38-support

dcdenu4 committed:
setting gdal version that effects precision for tests

dcdenu4 committed:
I believe this is removal of trailing whitespace

dcdenu4 committed:
better making of poly boxes and pep fixes

dcdenu4 commented:
@richpsharp @phargogh Just went through and addressed all the changes. Note that Ubuntu tests are still failing, but they work on Mac and Windows! I think that and the fact that I can get them to work using **setup-miniconda** (mentioned above) (but `windows-latest` fails using that), hopefully is enough to ignore that build/test issue for now and that it should be addressed in Issue #44. 

Let me know what else should be done for this PR! Thanks for hanging in there with me as I get back in the groove!

richpsharp commented on requirements.txt:
```diff
@@ -3,7 +3,7 @@
 # pygeoprocessing to work as expected.
 
-GDAL>=2.2.0,<3.0
+GDAL==3.0.4
```
I'm sorry I didn't notice this before, we've had problems in the past when we've pegged the version of packages to an *exact* version. Could you change this to `GDAL>=3.0`? Unless there's something in a particular version that you don't want before 3.0.4? Otherwise we'll have a pain if we have a docker image with 3.0.2 that'll need a whole rebuild etc.etc..

dcdenu4 replied:
Hey @richpsharp , There were some bug fixes since GDAL 3.0.2 that have effected the precision of transforms, therefore I was using 3.0.4 because that is what I created my tests in, and at least one of them would otherwise fail in 3.0.2 by a decimal. Let me know if you'd like to talk about this more.
I made the change to `GDAL>=3.0.4` in [83bd15](https://github.com/natcap/pygeoprocessing/pull/31/commits/83bd1546903dfef0eb234fdcb740cb551261e82d)

richpsharp replied:
Hi @dcdenu4, unless you think it needs to be 3.0.4 could you change it to `GDAL>=3.0.2`? (I've got a docker image that's using 3.0.2 that'd be a pain to rebuild, but I could if it was necessary).

richpsharp reviewed (changes requested):
Hi Doug, I'm sorry I had just one more thing on there which is the pegging of version 3.0.4 of GDAL in `requirements.txt`. If that exact version is not necessary can you loosen the requirements to >=3.0?

dcdenu4 committed:
Changing GDAL version to gte 3.0.4

davemfish reviewed (approved):
I took a look at these changes and don't have anything new to add.  Full disclosure: I didn't review the tests too closely.

dcdenu4 commented:
@richpsharp @phargogh We decided to leave the GDAL version at 3.0.4 via slack because of the mentioned bug fixes that GDAL has released. Therefore, requesting a re-review.

phargogh commented on tests/test_geoprocessing.py:
```diff
@@ -248,26 +249,37 @@ def test_reclassify_raster_empty_value_map(self):
+            443723.1273278, 4956276.905980, 443993.1273278, 4956546.905980]
+
+        polygon_a = shapely.geometry.box(
+            extents[0], extents[1], extents[2], extents[3])
```
FYI, python's argument unpacking to even further abbreviate this: `polygon_a = shapely.geometry.box(*extents)`.  What you have here will do exactly the same thing!

dcdenu4 replied:
You know I've seen you guys use this and thought about it, but must not have thought about it long enough to implement it. I will update it so that it helps me practice it.

dcdenu4 replied:
Done in [0f5e91](https://github.com/natcap/pygeoprocessing/pull/31/commits/0f5e91c7e1856f9ff8055b4bd8cff1516c412d99)

phargogh reviewed (approved):
Looks good to me, Doug!  I noted one other thing about the use of `shapely.geometry.box`, but that's more of an FYI than anything else.  Up to you if you'd like to change that.

dcdenu4 committed:
Issue #16 condensing shapely.geom.box call
